### PR TITLE
[helm] Add a loki canary test to the helm chart

### LINKED ISSUE #7228: [helm] Use Loki Canary in Helm Chart Test
**Is your feature request related to a problem? Please describe.**
When installing the helm chart, how do I know that it's not only installed semantically correct (ie. k8s pods are up and running), but also that Loki is behaving correctly?

**Describe the solution you'd like**
Include  a helm [chart test](https://helm.sh/docs/topics/chart_tests/) that utilizes the included Loki canary to test if Loki is running properly.

**Describe alternatives you've considered**
Just use the canary dashboard, a slightly more manual process.

### PR DESCRIPTION
**Pull request by trevorwhitney**
**What this PR does / why we need it**:

Add a helm chart test that utilizes the Loki canary to confirm the Loki cluster is functioning correctly.

**Which issue(s) this PR fixes**:
Fixes #7228

### CODE REVIEW

File: .drone/drone.jsonnet [hunk 1]
```diff
@@ -480,6 +480,36 @@ local manifest_ecr(apps, archs) = pipeline('manifest-ecr') {
       },
     ],
   },
+  pipeline('loki-canary-test-image') {
```
Comment (jeschkies) [inline .drone/drone.jsonnet:483]: Hm, the name might be misleading.

File: .github/workflows/helm-chart-test.yaml [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+          kubectl --namespace prometheus get services -l "release=prometheus"
+
+      - name: Run chart-testing (install)
+        run: |
+          changed=$(ct list-changed --config "${CT_CONFIGFILE}")
+          if [[ "$changed" == "charts/enterprise-metrics" ]]; then
+            # Do not run `ct install` for enterprise-metrics
+            exit 0
+          fi
+          ct install --config "${CT_CONFIGFILE}"
```
Comment (jeschkies) [inline .github/workflows/helm-chart-test.yaml:86]: No really sure how this works. How is this running the test image?

File: cmd/loki-canary-test/README.md [hunk 3]
```diff
@@ -0,0 +1,7 @@
+# Loki Canary Test
+
+Loki canary test is a tool to verify that a loki canary is running correctly, and as a result, a good way to test that a Loki cluster with a canary deployed is also running correctly. It's primary use is for testing the helm chart after installation, but can be used standalone to test any Loki Canary / Cluster.
```
Comment (jeschkies) [inline cmd/loki-canary-test/README.md:3]:
```suggestion
Loki canary test is a tool to verify that a Loki canary is running correctly, and as a result, a good way to test that a Loki cluster with a canary deployed is also running correctly. It's primary use is for testing the Helm chart after installation, but can be used standalone to test any Loki Canary / Cluster.
```

File: cmd/loki-canary-test/canary/client_test.go [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+	"github.com/grafana/loki/cmd/loki-canary-test/canary"
+	v1 "github.com/prometheus/client_golang/api/prometheus/v1"
+	"github.com/prometheus/common/model"
+	"github.com/stretchr/testify/assert"
+)
+
+const sumEntriesQuery = "sum(loki_canary_entries_total)"
+const sumEntriesMissingQuery = "sum(loki_canary_missing_entries_total)"
+
+func TestCanary(t *testing.T) {
```
Comment (jeschkies) [inline cmd/loki-canary-test/canary/client_test.go:18]: This is not run as part of the main is it?

File: production/helm/loki/ci/default-values.yaml [hunk 5]
```diff
@@ -6,3 +6,11 @@ read:
...
 write:
   replicas: 1
+monitoring:
+  serviceMonitor:
+    labels:
+      release: "prometheus"
+  selfMonitoring:
+    lokiCanary:
+      test:
+        prometheusAddress: "http://prometheus-kube-prometheus-prometheus.prometheus.svc.cluster.local:9090"
```
Comment (jeschkies) [inline production/helm/loki/ci/default-values.yaml:16]: Is this an option for the Canary?

File: cmd/loki-canary-test/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,12 @@
+FROM golang:1.18.5 as build
```
Comment (jeschkies) [inline cmd/loki-canary-test/Dockerfile:1]: I'm not sure we want to publish this. As users should use alerts on canary metrics instead.

**Review by jeschkies — CHANGES_REQUESTED** [review]
Thanks for this awesome work.

What do you think about running `go test` instead?

File: cmd/loki-canary-test/Dockerfile (hunk 6 above)
Comment (trevorwhitney) [reply cmd/loki-canary-test/Dockerfile]: We need to publish these tests in a way that is easily consumable by a k8s Pod spec, so I think publishing an image with the compiled test binary is probably the easiest way to do that.

File: production/helm/loki/ci/default-values.yaml (hunk 5 above)
Comment (trevorwhitney) [reply production/helm/loki/ci/default-values.yaml:16]: For the test. It needs to know which prometheus to query the canary's metrics from.

File: cmd/loki-canary-test/canary/client_test.go (hunk 4 above)
Comment (trevorwhitney) [reply cmd/loki-canary-test/canary/client_test.go]: switch to not having main and running everything as a `go test`

**Comment by trevorwhitney** [comment]
@jeschkies changed this to use `go test`, changed the image name, and moved the source into the helm folder. Ready for another review, thanks!

File: Makefile [hunk 7]
```diff
@@ -149,6 +149,15 @@ loki-canary: cmd/loki-canary/loki-canary
 cmd/loki-canary/loki-canary:
 	CGO_ENABLED=0 go build $(GO_FLAGS) -o $@ ./$(@D)
 
+###############
+# Helm-Test #
+###############
+.PHONY: production/helm/loki/src/helm-test/helm-test
+helm-test: production/helm/loki/src/helm-test/helm-test
+
+production/helm/loki/src/helm-test/helm-test:
```
Comment (jeschkies) [inline Makefile:158]:
```suggestion
# Package Helm tests but do not run them.
production/helm/loki/src/helm-test/helm-test:
```

File: production/helm/loki/src/helm-test/README.md [hunk 8]
```diff
@@ -0,0 +1,7 @@
+# Loki Helm Test
+
+This folder contains a collection of go tests that test if a Loki canary is running correctly. It's primary use it to test that the helm chart is working correctly by using metrics from the Loki canary. In the helm chart, the template for this test is only available if you are running both the Loki canary and have self monitoring enabled (as the Loki canary's logs need to be in Loki for it to work). However, the tests in this folder can be run against any running Loki canary using `go test`.
```
Comment (jeschkies) [inline production/helm/loki/src/helm-test/README.md:3]: Nice!

File: production/helm/loki/src/helm-test/canary_test.go [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+
+		client, err := api.NewClient(api.Config{
+			Address: address,
+		})
+		require.NoError(t, err, "Failed to create Loki Canary client")
+
+		return v1.NewAPI(client)
+	}
+
+	withRetries := func(t *testing.T, test func() error) error {
```
Comment (jeschkies) [inline production/helm/loki/src/helm-test/canary_test.go:56]: What do you think about using [Eventuallyf](https://pkg.go.dev/github.com/stretchr/testify/require#Eventually) instead?

**Review by jeschkies — CHANGES_REQUESTED** [review]
This is really awesome. Nice work! I have only a few points.

File: .drone/drone.jsonnet [hunk 10]
```diff
@@ -480,6 +480,36 @@ local manifest_ecr(apps, archs) = pipeline('manifest-ecr') {
...
+        settings: {
+          repo: 'grafana/loki-helm-test',
+          dockerfile: 'production/helm/loki/src/helm-test/Dockerfile',
+          dry_run: true,
+        },
+      },
+      {
+        name: 'push-image',
+        image: 'plugins/docker',
+        when: onTagOrMain + onPath('production/helm/loki/src/helm-test/**'),
```
[the remark this answers is not in the document]
Comment (trevorwhitney) [reply .drone/drone.jsonnet:502]: Correct, as the new image needs to be published before we can use it in test. We could push an image per-build, but we'd have to also override the helm configuration to tell the test pod to use that image.

File: production/helm/loki/src/helm-test/canary_test.go (hunk 9 above)
Comment (trevorwhitney) [reply production/helm/loki/src/helm-test/canary_test.go:56]: sure

**Comment by trevorwhitney** [comment]
@jeschkies finally got a chance to take a look at this and make some of the changes you requested, let me know what you think.

File: .drone/drone.jsonnet (hunk 10 above)
[the remark this answers is not in the document]
Comment (jeschkies) [reply .drone/drone.jsonnet:502]: It's fine. We have a similar issue with the build image.

File: production/helm/loki/src/helm-test/Dockerfile [hunk 11]
```diff
@@ -0,0 +1,10 @@
+FROM golang:1.18.5 as build
+
+COPY . /src/loki
```
Comment (jeschkies) [inline production/helm/loki/src/helm-test/Dockerfile:3]: So docker build must be called from the root of the repository, right?

File: production/helm/loki/src/helm-test/README.md [hunk 12]
```diff
@@ -0,0 +1,7 @@
...
+
+## Instructions
+
+Run `go test .` from this directory, or use the docker image published at `grafana/loki-helm-test`.
```
Comment (jeschkies) [inline production/helm/loki/src/helm-test/README.md:7]:
```suggestion
Run `go test .` from this directory, or use the Docker image published at `grafana/loki-helm-test`.
```

File: production/helm/loki/src/helm-test/canary_test.go [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+	if value, ok := os.LookupEnv(key); ok {
+		return value
+	}
+	return fallback
+}
+
+func testResult(result model.Value, query string, test func(model.SampleValue) bool, msg string) error {
+	if v, ok := result.(model.Vector); ok {
+		for _, s := range v {
+			fmt.Printf("%s => %v\n", query, s.Value)
```
Comment (jeschkies) [inline production/helm/loki/src/helm-test/canary_test.go:27]: What do you think about using `t.Log` instead?

File: production/helm/loki/src/helm-test/canary_test.go [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+
+	eventually := func(t *testing.T, test func() error, msg string) {
+		timeout := getEnv("CANARY_TEST_TIMEOUT", "1m")
+		timeoutDuration, err := time.ParseDuration(timeout)
+		require.NoError(t, err, "Failed to parse timeout. Please set CANARY_TEST_TIMEOUT to a valid duration.")
+
+		require.Eventually(t, func() bool {
+			queryError := test()
+			if queryError != nil {
+				fmt.Printf("Query failed\n%+v\n", queryError)
```
Comment (jeschkies) [inline production/helm/loki/src/helm-test/canary_test.go:63]:
```suggestion
				t.Logf("Query failed\n%+v\n", queryError)
```

File: production/helm/loki/src/helm-test/canary_test.go [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+	}
+
+	return fmt.Errorf("unexpected Prometheus result type: %v ", result.Type())
+}
+
+func TestCanary(t *testing.T) {
+	totalEntriesQuery := "sum(loki_canary_entries_total)"
+	totalEntriesMissingQuery := "sum(loki_canary_missing_entries_total)"
+
+	testContext := func(t *testing.T) v1.API {
```
Comment (jeschkies) [inline production/helm/loki/src/helm-test/canary_test.go:43]:
```suggestion
	newClient := func(t *testing.T) v1.API {
```

Actually, this method could be moved out of the test function.

File: production/helm/loki/src/helm-test/canary_test.go [hunk 16]
```diff
@@ -0,0 +1,102 @@
...
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+
+	t.Cleanup(func() {
+		cancel()
+	})
+
+	t.Run("Canary should have entries", func(t *testing.T) {
+		client := testContext(t)
```
Comment (jeschkies) [inline production/helm/loki/src/helm-test/canary_test.go:76]:
```suggestion
		client := newClient(t)
```

File: production/helm/loki/src/helm-test/canary_test.go [hunk 17]
```diff
@@ -0,0 +1,102 @@
...
+				return err
+			}
+			return testResult(result, totalEntriesQuery, func(v model.SampleValue) bool {
+				return v > 0
+			}, fmt.Sprintf("Expected %s to be greater than 0", totalEntriesQuery))
+		}, "Expected Loki Canary to have entries")
+	})
+
+	t.Run("Canary should not have missed any entries", func(t *testing.T) {
+		client := testContext(t)
```
Comment (jeschkies) [inline production/helm/loki/src/helm-test/canary_test.go:90]:
```suggestion
		client := newClient(t)
```

File: production/helm/loki/values.yaml [hunk 18]
```diff
@@ -529,6 +529,28 @@ rbac:
   # -- For OpenShift set pspEnabled to 'false' and sccEnabled to 'true' to use the SecurityContextConstraints.
   sccEnabled: false
 
+# -- Section for configuring optional helm test
```
Comment (jeschkies) [inline production/helm/loki/values.yaml:532]:
```suggestion
# -- Section for configuring optional Helm test
```

**Review by jeschkies — APPROVED** [review]
Really nice! I have only a few nits.

I did not review the flake changes as I'm not an expert.

File: tools/dev/k3d/environments/enterprise-logs/spec.json [hunk 19]
```diff
@@ -6,7 +6,7 @@
     "namespace": "environments/enterprise-logs/main.jsonnet"
   },
   "spec": {
-    "apiServer": "https://0.0.0.0:44365",
+    "apiServer": "https://0.0.0.0:45985",
```
[the remark this answers is not in the document]
Comment (trevorwhitney) [reply tools/dev/k3d/environments/enterprise-logs/spec.json:9]: yeah, just one of those things that will change for everyone, but we can't gitignore :(

**Pull request merged**